Documentation/ServicingGranularity.md

```diff
@@ -0,0 +1,146 @@
+of versions of a specific repo is bounded by the number of products it builds.
+The max is only three for Core-Setup, but larger (dozens?) for CoreFX because of
+the number of independent OOB packages.
+
+## Consequences of depending on multiple versions of the same repo
+
+### Microsoft builds
+This is completely fine for Microsoft servicing builds, because they use
+prebuilts. When building a repo, each upstream product is downloaded from an
+online source, so it doesn't matter which version of the product is being pulled
+down.
+
+### Building from source
+This is bad for source-builds, where downloading prebuilt products is forbidden.
+Every upstream product must be built from source.
+
+## Resolution: build all dependency nodes
+
+The obvious fix is to build every dependency node from source: multiple commits
+of Core-Setup and CoreFX (and others). However, this places a burden on a lot of
+people.
+
+The number of dependency nodes will tend to increase over time as the branch
+gets serviced, as demonstrated above. That means the difficulty and complexity
+of building .NET Core would then depend on how many times the branch has had a
+servicing release.
+
+A few concrete reasons this has significant overhead and/or upfront work for
+source-build infra maintainers:
+
+1. Say the world has changed in a way that makes it impossible to build anything
+   earlier than CoreFX `v3.0.3` on a new distro or platform. Then, patches need
+   to be applied to `v3.0.0`, `v3.0.1`, and `v3.0.2` to get them to build,
+   because the OOB packages are required. The patching work *probably* only
+   needs to be done once and applied on all affected commits, but as the release
+   branch continues to be serviced, the amount of work to get the product built
+   on a new distro balloons.
+2. Each repo must be checked out at every required commit/tag, which syncs a
+   *lot* of irrelevant source code.
+3. The repo must either be built in its entirety, or something needs to be
+   figured out to allow source-build to pick the exact product it needs to
+   build.
+
+### Can we split up source-build products to make this less difficult?
+At first glance, it looks like source-build should simply be changed to treat
+products the same as the Microsoft build does.
+
+> Currently, source-build is set up to build one ".NET Core SDK + Runtime"
+product, which is then split into runtime and SDK packages. The build process
+has no concept of the way Microsoft services what it considers to be the
+products. This means everything that goes into the SDK product, like targeting
+packs and OOB corefx packages, needs to be built all at the same time. Artifacts
+from a previously source-built SDK can't be redistributed for use by the new
+SDK.
+
+There is feedback that we need to keep it simple, continue building a ".NET Core
+SDK + Runtime" product. Asking distro maintainers to keep track of many products
+is a significant burden.
```

Comment:
+1. I dont think it's a good idea to have a complex process with the expectation that everyone gets it right.